man/smsSynth.1
.TH smsSynth 1 "2008 Feb 22" GNU
.SH NAME
smsSynth - SMS Synthesis command line application.
.SH SYNOPSIS
.B smsSynth
[-\fIoption\fP ...]
.I inputSoundFile outputSmsFile
.SH DESCRIPTION
\fISMS\fP is a set of techniques and software implementations for the
analysis, transformation and synthesis of musical sounds based on a
sinusoidal plus residual model. These techniques can be used for
synthesis, processing and coding applications, while some of the
intermediate results might also be applied to other music related
problems, such as sound source separation, musical acoustics, music
perception, or performance analysis. The basic model and
implementation was developed in the PhD thesis by X. Serra in 1989 and
since then many extensions have been proposed at MTG-UPF and by other
researchers.

\fIsmsSynth\fP will synthesize a .sms file analyzed wish smsAnal.  By default,
the synthesis will contain deterministic and stochastic components added 
together, but you can optionally choose to just synthesize one or the other. 
For background information on how to create the analysis file, or what to do with 
it other than recreate the original, see the README included with the SMS package
or visit the SMS homepage at:
\fIhttp://www.iua.upf.es/~sms/\fP
.SH OPTIONS

.TP 8
.BI -v " verbose mode"
Print out lots of information about parameters and what is going on.
.TP 8
.BI -r " samplerate"
.B (default: original sound file's samplerate)
output sampling rate.
.TP 8
.BI -s " synth-type"
.B (default: 0) [0,1,2]
Synthesis type. 0: all, 1: deterministic only, 2: stochastic only
.TP 8
.BI -d " det-synth-type"
.B (default: 0) [0,1,2]
method of deterministic synthesis type. 0: IFFT (defailt), 1: oscillator bank
.TP 8
.BI -h " hop"
.B (default: 128)
Hop size in samples from one frame to the next. 128 <= sizeHop <= 8092, rounded to a power of 2
.TP 8
.BI -t " time-factor"
.B (default: 1)
Positive value to scale by overall time.
.TP 8
.BI -g " stoc-gain"
.B (default: 1)
Positive value to multiply into stochastic gain.
.TP 8
.BI -x " transpose"
.B (default: 0)
Transpose value based on the Equal Tempered Scale.
.TP 8
.BI -i " interp"
.B (default: 1, on)
Interpolate between frames when time scaling. 
.TP 8
.BI -f " file-type"
.B (default: 0)
Output soundfile type (default 0): 0 is wav, 1 is aiff
.SH SEE ALSO
smsAnal(1), smsClean(1), smsPrint(1), smsResample(1)



doc/changes.txt
=-=-=-=-=-=-=- libsms change log -=-=-=-=-=-=

=== June 23, 2009 === version 1.1 beta =====================================================

New features:
	- Can now transform sounds using their spectral envelopes. By setting the appropriate
	flags in an SMS_ModifyParams struct and calling sms_modify, it is possible to shape the
	partials of one sound using the spectral envelope of another (or a linear interpolation
	anywhere between their two envelopes), and to transpose sounds while maintaining the
	original spectral envelope. In fact the envelopes do not have to be related to any
	analysed sounds, they could be generated by any means and then applied to the detected
	partials of a given sound, so purely synthetic envelopes could be used as well.
	- The full SMS analysis/synthesis process is now callable from Python, and a Python
	module called SMS.py has been created that will analyze or synthesize a file with a single
	function call. See the demo script python/AnalysisSynthesis.py for an example of using it.
        - new misc. functions: sms_rms, sms_arrayScalarTempered

API changes:
	- All transformations (including transpostion and stochastic component scaling) should
	now be performed using the sms_modify function. The required transformations are specified
	by using the SMS_ModifyParams structure. Examples of using this (from Python) can be found
	in the file python/modify_examples.py.
        - dB conversion is slightly different.  It is normalized to above 0, using a threshhold (default 0.00001)
          to avoid NaN's.  You can set this with sms_setMagThresh
        - all magnitudes are now stored in linear
        - windows are now left un-normalized by default. Call sms_scaleWindow() to normalize it.
Bug fixes:
	- The stochastic gain calculated during residual analysis was being applied twice during
	resynthesis, fixed this.
	- Also changed residual and spectrum energy calculation in sms_residual so that they use 
	the sum of the square of signal samples rather than just the sum of the samples.
	- Fixed a call to memset is sms_synthesize that was writing more data than intended (writing 
	more than the given array size)
	- made some changes to python/setup.py so that the python module builds on OS X (tested
	on Leopard).

Other:
	- Added some unit tests (in Python) to test/Tests.py
        - smsToYaml is gone. smsPrint now naturally prints in yaml format, and you can save it to file.

=== April 9, 2009 === version 1.05e, experimental release =====================================
This is a release heading in the direction of better modularity, especially in the analysis routines.  The ultimate
goal is to be able to interchange each step of the analysis process (such as f0 estiumation, transient extraction,
etc) with custom code, either in C or python.

Many functions received argument changes so they could be easily wrapped in python using SWG.  I tried to
give every array pointer an argument before it that indicates the size of the array.  In some functions, this seemed
too redundant (such as sms_spectrum for example), so they are custom wrapped in the swig interface file.

In doing this, the release is made incompatible with the last because of many name and argument changes. This will
probably continue to happen through the summer.  If projects break or people think things were better the way before,
please email me.

Notable changes:
        - some files have been moved around in an attempt to organize function types (transforms.c. filters.c, tables.c, etc)
        - some more general functions are in sms.c, such as RMS array calculation and frequency conversion (well-tempored)
        - started replacing float with sfloat so that, one day, the library can be optionally compiled in double precision
        - added support for reading of multi-channel audio files - can select what channel you want to analyze 
        - smsAnal now depends much more on library functions (like sms_initAnalParams)
        - tried to fix documentation errors mentioned to me
        - removed some outdated tools (smsToLisp, smsSynthDet - now in smsSynth)
        - moved around synthesis buffers so that they are properly re-allocated if the sizeHop changes
        - using Mersainne Twister random number generator.
        - sms_analyzeFrame() is supposed to be a function for computing sinusoidal analysis without a circular buffer.
          Its not quite at that point yet.
        - SMS_PeakParams structure has parameters just for sinusoidal analysis (it used to need all of SMS_AnalParams)
        - python test scripts have been added for many things (all names test_*.py)

=== Nov. 15, 2008 === version 1.0, first release ============================================

This release represents a revival of the original code from Xavier Serra, which dates back to the early 1990's.
Care has been taken to make sure the functionality of the algorithms important to SMS have not changed,
while still updating function calls to work with modern day platforms.  Future releases will focus more on incorporating
improvements in the SMS model that have since been discovered. Some changes to the original code include:
        - using libsndfile for sound file handling
        - using OOURA fft routines for real-DFT functions (fftw3 as well, but not fully working)
        - optionally can compile to use FFTW3 fft libary, but it is still in a testing stage and fails in pd externals
        - appended the namespaces sms_* and SMS_ to all globally seen functions, structures and defines/enumerations
        - uses scons instead of make, as it is very powerful, being based in python
        - python SWIG wrappers
        - synthesis parameters "stocgain" and "ftranspose" are added to change the current synthesis, but not the data itself
        - smsToYaml - converts a binary *.sms file to YAML text format, which is very easy to read

tools/smsSynth.c
/* 
 * Copyright (c) 2008 MUSIC TECHNOLOGY GROUP (MTG)
 *                         UNIVERSITAT POMPEU FABRA 
 * 
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 */
/*
 *
 *    smsSynth - program for synthesizing an sms file from the command line.
 *
 */

#include "sms.h"
#include <popt.h>

const char *help_header_text =
        "\n\n"
        "Usage: smsSynth [options]  <inputSmsFile> <outputSoundFile>\n"
        "\n\n"
        "synthesize an analysis (.sms) file made with smsAnal. "
        "output file format is 32bit floating-point WAV or AIFF."
        "\n\n";


int main (int argc, const char *argv[])
{
	char *pChInputSmsFile = NULL, *pChOutputSoundFile = NULL;
	SMS_Header *pSmsHeader = NULL;
	FILE *pSmsFile; /* pointer to sms file to be synthesized */
	SMS_Data smsFrameL, smsFrameR, smsFrame; /* left, right, and interpolated frames */
	float *pFSynthesis; /* waveform synthesis buffer */
	long iSample, i, nSamples, iLeftFrame, iRightFrame;
        float fFrameLoc; /* exact sms frame location, used to interpolate smsFrame */
        float fFsRatio,  fLocIncr;
        int verbose = 0;
        int iSoundFileType = 0; /* wav file */
        int doInterp = 1;
        float timeFactor = 1.0;
	SMS_SynthParams synthParams;
        sms_initSynthParams(&synthParams); /* set some default params that may be updated */

	int optc;   /* switch */
	poptContext pc;
        struct poptOption	options[] =
                {
                        {"verbose", 'v', POPT_ARG_NONE, &verbose, 0, 
                         "verbose mode", 0},
                        {"samplerate", 'r', POPT_ARG_INT, &synthParams.iSamplingRate, 0, 
                         "sampling rate of output sound (default is original)", "int"},
                        {"synth-type", 's', POPT_ARG_INT, &synthParams.iSynthesisType, 0, 
                         "synthesis type (0: all (default), 1: deterministic only , 2: stochastic only)", "int"},
                        {"det-synth-type", 'd', POPT_ARG_INT, &synthParams.iDetSynthType, 0, 
                         "method of deterministic synthesis type (0: IFFT (default) , 1: oscillator bank)", "int"},
                        {"hop", 'h', POPT_ARG_INT, &synthParams.sizeHop, 0, 
                         "sizeHop (default 128) 128 <= sizeHop <= 8092, rounded to a power of 2", "int"},
                        {"time-factor", 't', POPT_ARG_FLOAT, &timeFactor, 0, 
                         "time factor (default 1): positive value to scale by overall time", "float"},
                        {"stoc-gain", 'g', POPT_ARG_FLOAT, &synthParams.fStocGain, 0, 
                         "stochastic gain (default 1): positive value to multiply into stochastic gain", "float"},
                        {"transpose", 'x', POPT_ARG_FLOAT, &synthParams.modParams.transpose, 0, 
                         "transpose factor (default 0): value based on the Equal Tempered Scale", "float"},
                        {"interp", 'i', POPT_ARG_INT, &doInterp, 0, 
                         "interpolate between frames when time scaling (default on, 0=off)", "int"},
                        {"file-type", 'f', POPT_ARG_INT, &iSoundFileType, 0, 
                         "output soundfile type (default 0): 0 is wav, 1 is aiff", "int"},
                        POPT_AUTOHELP
                        POPT_TABLEEND
                };

        pc = poptGetContext("smsSynth", argc, argv, options, 0);
        poptSetOtherOptionHelp(pc, help_header_text);

        if (argc <= 1)
        {
                poptPrintUsage(pc,stderr,0);
                return 1;
        }

        while ((optc = poptGetNextOpt(pc)) > 0) {
/*                 switch (optc) { */
/*                         /\* specific arguments are handled here *\/ */
/*                 } */
        }
        if (optc < -1) 
        {
                /* an error occurred during option processing */
                printf("%s: %s\n",
                       poptBadOption(pc, POPT_BADOPTION_NOALIAS),
                       poptStrerror(optc));
                return 1;
        }

	pChInputSmsFile = (char *) poptGetArg(pc);
	pChOutputSoundFile = (char *) poptGetArg(pc);
        /* parsing done */
        
        sms_getHeader (pChInputSmsFile, &pSmsHeader, &pSmsFile);

	if (sms_errorCheck())
	{
                printf("error in sms_getHeader: %s", sms_errorString());
                exit(EXIT_FAILURE);
	}	    

        sms_init();
        sms_initSynth( pSmsHeader, &synthParams );
        /* disable interpolation for residual resynthesis with original phases (not implemented yet) */
        if(pSmsHeader->iStochasticType == SMS_STOC_IFFT)
                doInterp = 0;

        synthParams.modParams.doTranspose = 1; /* turns on transposing (whether there is a value or not */

        if(verbose)
        {
                printf("__arguments__\n");
                printf("samplingrate: %d \nsynthesis type: ", synthParams.iSamplingRate);
                printf(" do frame interpolation: %d \n", doInterp);
                if(synthParams.iSynthesisType == SMS_STYPE_ALL) printf("all ");
                else if(synthParams.iSynthesisType == SMS_STYPE_DET) printf("deterministic only ");
                else if(synthParams.iSynthesisType == SMS_STYPE_STOC) printf("stochastic only ");
                printf("\ndeteministic synthesis method: ");
                if(synthParams.iDetSynthType == SMS_DET_IFFT) printf("ifft ");
                else if(synthParams.iDetSynthType == SMS_DET_SIN) printf("oscillator bank ");
                printf("\nsizeHop: %d \n", synthParams.sizeHop);
                printf("time factor: %f \n", timeFactor);
                printf("stochastic gain factor: %f \n", synthParams.fStocGain);
                printf("frequency transpose factor: %f \n", synthParams.modParams.transpose);
                printf("__header info__\n");
                printf("original samplingrate: %d, iFrameRate: %d, origSizeHop: %d\n", pSmsHeader->iSamplingRate,
                       pSmsHeader->iFrameRate, synthParams.origSizeHop);
                printf("original file length: %f seconds \n",
                       pSmsHeader->nFrames / (float) pSmsHeader->iFrameRate);
                if(!iSoundFileType) printf("output soundfile type: wav \n");
                else  printf("output soundfile type: aiff \n");
        }

        /* initialize libsndfile for writing a soundfile */
	sms_createSF ( pChOutputSoundFile, synthParams.iSamplingRate, iSoundFileType);

	/* setup for synthesis from file */
        if(doInterp)
        {
                sms_allocFrameH (pSmsHeader, &smsFrameL);
                sms_allocFrameH (pSmsHeader, &smsFrameR);
        }
        sms_allocFrameH (pSmsHeader, &smsFrame); /* the actual frame to be handed to synthesizer */

	if ((pFSynthesis = (float *) calloc(synthParams.sizeHop, sizeof(float)))
	    == NULL)
        {
                printf ("Could not allocate memory for pFSynthesis");
                exit(0);
        }

	iSample = 0;
        /* number of samples is a factor of the ratio of samplerates */
        /* multiply by timeFactor to increase samples to desired file length */
        /* pSmsHeader->iSamplingRate is from original analysis */
        fFsRatio = (float) synthParams.iSamplingRate / pSmsHeader->iSamplingRate;
	nSamples = pSmsHeader->nFrames * synthParams.origSizeHop * timeFactor * fFsRatio;

        /* divide timeFactor out to get the correct frame */
        fLocIncr = pSmsHeader->iSamplingRate / 
                ( synthParams.origSizeHop * synthParams.iSamplingRate * timeFactor); 

	while (iSample < nSamples)
	{
		if(doInterp)
                {
                        fFrameLoc =  iSample *  fLocIncr;
                        /* left and right frames around location, gaurding for end of file */
                        iLeftFrame = MIN (pSmsHeader->nFrames - 1, floor (fFrameLoc)); 
                        iRightFrame = (iLeftFrame < pSmsHeader->nFrames - 2)
                                ? (1+ iLeftFrame) : iLeftFrame;
                        sms_getFrame (pSmsFile, pSmsHeader, iLeftFrame, &smsFrameL);
                        sms_getFrame (pSmsFile, pSmsHeader, iRightFrame,&smsFrameR);
                        sms_interpolateFrames (&smsFrameL, &smsFrameR, &smsFrame,
                                               fFrameLoc - iLeftFrame);
                }
                else
                {
                        sms_getFrame (pSmsFile, pSmsHeader, (int) iSample * fLocIncr, &smsFrame);
                        printf("frame: %d \n",  (int) (iSample * fLocIncr));
                }
                sms_modify(&smsFrame, &synthParams.modParams); 
                sms_synthesize (&smsFrame, pFSynthesis, &synthParams);
		sms_writeSound (pFSynthesis, synthParams.sizeHop);
    
		iSample += synthParams.sizeHop;

                if(verbose)
                {
                        if (iSample % (synthParams.sizeHop * 20) == 0)
                                fprintf(stderr,"%.2f ", iSample / (float) synthParams.iSamplingRate);
                }

	}

        if(verbose)
        {
                printf("\nfile length: %f seconds\n", (float) iSample / synthParams.iSamplingRate);
        }
        printf("wrote %ld samples in %s\n",  iSample, pChOutputSoundFile);

	/* close output sound file, free memory and exit */
	sms_writeSF ();
        if(doInterp)
        {
                sms_freeFrame(&smsFrameL);
                sms_freeFrame(&smsFrameR);
        }
        sms_freeFrame(&smsFrame);
	free (pFSynthesis);
        free (pSmsHeader);
        sms_freeSynth(&synthParams);
        sms_free();
	return(1);
}
